File: user-service/utils/kafka_producer.py
from kafka import KafkaProducer
import json
import os
from dotenv import load_dotenv

load_dotenv()

KAFKA_BROKER = os.getenv("KAFKA_BROKER")
KAFKA_TOPIC = os.getenv("KAFKA_TOPIC")

# Initialize producer
producer = KafkaProducer(
    bootstrap_servers=[KAFKA_BROKER],
    value_serializer=lambda v: json.dumps(v).encode('utf-8')
)

def send_notification_event(email, subject, message):
    """Publish notification message to Kafka topic"""
    event_data = {
        "email": email,
        "subject": subject,
        "message": message
    }
    producer.send(KAFKA_TOPIC, event_data)
    producer.flush()
    print(f"📤 Sent event to Kafka: {event_data}")

def send_otp_email(email, otp, name=None):
    otp_data = {
        email,
        "Your OTP Verification Code",
        f"Hello{(' ' + name) if name else ''},\n\nYour OTP code is: {otp}\nIt will expire in a few minutes.\n\nIf you didn't request this, ignore this email."
    }
    producer.send(KAFKA_BROKER, otp_data)
    producer.flush()
    print(f"📤 Send otp to kafka: {otp_data}")
    

File: event-registration-service/utils/kafkaProducer.js
const { Kafka } = require('kafkajs');

const kafka = new Kafka({
  clientId: 'event-registration-service',
  brokers: [process.env.KAFKA_BROKER],
});

const producer = kafka.producer();

async function initKafkaProducer() {
  await producer.connect();
  console.log('✅ Kafka Producer connected');
}

async function sendNotification(email, subject, message) {
  try {
    data = {
        "email": email,
        "subject": subject,
        "message": message
    }

    await producer.send({
      topic: process.env.KAFKA_TOPIC,
      messages: [{ value: JSON.stringify(data) }],
    });
    console.log(`📤 Notification sent: ${JSON.stringify(data)}`);
  } catch (error) {
    console.error('❌ Error sending Kafka message:', error);
  }
}

module.exports = { initKafkaProducer, sendNotification };
